File: server/lib/battle/Crit.js
'use strict';

/*
 * Gen1 critical hits.
 *
 * Node half of src/BattleSim's crit step; same parity contract as Damage.js.
 *
 * Two things here are bugs that have to be kept:
 *
 *   * Focus Energy *quarters* the crit threshold instead of quartering the
 *     odds of not critting. It makes crits rarer. The generation shipped that
 *     way, every reimplementation reproduces it, and a hub that helpfully
 *     inverted it would hand one side of a mediated battle a buff the other
 *     side's client never agreed to.
 *   * The threshold is compared against a 256-value roll but clamps at 255, so
 *     even a high-crit-ratio move with a fast user cannot crit on roll 255.
 *
 * Order matters: Focus Energy divides first, the high-ratio multiplier
 * applies after, and only the final value is clamped -- so the two stack to
 * 96 rather than saturating at 255 and then being quartered to 63.
 */

const MIN = 0;
const MAX = 255;
const HIGH_RATIO = 8;

const idiv = (a, b) => Math.floor(a / b);

const clamp = (v, lo, hi) => (v < lo ? lo : v > hi ? hi : v);

// opts: { focusEnergy, highCritMove }
function threshold(baseSpeed, opts) {
  const o = opts || {};
  let t = idiv(Math.max(0, baseSpeed || 0), 2);
  if (o.focusEnergy) t = idiv(t, 4);
  if (o.highCritMove) t = t * HIGH_RATIO;
  return clamp(t, MIN, MAX);
}

/*
 * check({ baseSpeed, focusEnergy, highCritMove, roll })
 *
 * `baseSpeed` is the species base speed, not the battler's current speed --
 * Gen1 crit rate ignores stat stages and paralysis alike.
 */
function check(input) {
  const t = threshold(input.baseSpeed, input);
  return { threshold: t, isCrit: input.roll < t };
}

module.exports = { check, threshold, MIN, MAX, HIGH_RATIO };

File: src/BattleSim/Crit.lua
-- Gen1 critical hits, including the Focus Energy bug.
--
--   threshold = floor(baseSpeed / 2)
--   Focus Energy:  threshold = floor(threshold / 4)
--   high-crit move: threshold = threshold * 8
--   clamp to 0..255, then isCrit = roll < threshold
--
-- Focus Energy *quarters* the threshold instead of doubling it.  That is the
-- original's famous inverted shift and it is kept, for the same reason the
-- 1/256 miss is kept in Accuracy.lua: this sim has to resolve a fight the way
-- the game the players are playing would resolve it, not the way the manual
-- says it should.  A mod that silently corrected it would make every mediated
-- battle diverge from every offline one.
--
-- The clamp gives the two ends their behaviour for free: a threshold of 0
-- never crits (no roll is below 0) and a threshold of 255 still fails on roll
-- 255, so nothing in Gen1 crits every single turn.
--
-- Note this is *base* speed, the species' constant -- not the battler's
-- current speed, which is why paralysis quartering speed does not lower the
-- crit rate.
--
-- Nothing here raises; no love, no engine modules, no mod facade.

local need = ...            -- entry-chunk shape; this module resolves nothing

local M = {}

local floor, max, min = math.floor, math.max, math.min

M.MIN = 0
M.MAX = 255
M.HIGH_RATIO = 8

local function int(value, fallback)
  local n = tonumber(value)
  if not n or n ~= n then return fallback end
  return floor(n)
end

-- opts: { focusEnergy, highCritMove }
function M.threshold(baseSpeed, opts)
  opts = opts or {}
  local t = floor(max(0, int(baseSpeed, 0)) / 2)
  if opts.focusEnergy then t = floor(t / 4) end
  if opts.highCritMove then t = t * M.HIGH_RATIO end
  return min(M.MAX, max(M.MIN, t))
end

-- Returns isCrit (boolean) and the threshold it was decided against, so a
-- caller logging a surprising crit can show the number without recomputing it.
function M.check(baseSpeed, roll, opts)
  local threshold = M.threshold(baseSpeed, opts)
  return int(roll, M.MAX) < threshold, threshold
end

return M
